### PR TITLE
refactor(core): useForm improvements

### LINKED ISSUE #5460: [BUG] useForm onFinishUpdate promise never settles when in pessimistic mutationMode
### Describe the bug

The useForm hook is used with mutationMode set to "pessimistic" and the update mutation from the data provider rejects.
The issue in the code is here:  https://github.com/refinedev/refine/blob/master/packages/core/src/hooks/form/useForm.ts#L581
In that code path, the promise executor returns without ever calling reject (nor resolve) so the promise is never settled. This means the onFinish function never rejects (nor resolves) and it also means the underlying react-hook-form state remains with `isSumitting` as true indefinitely.

### Steps To Reproduce

1. Implement a component with useForm configured with mutationMode set to "pessimistic"
2. Add a then and catch to the onFinish method with console log statements to tell when its promise resolves or rejects
3. Add a console log to log the isSubmitting form state on every render of the component
4. Submit a form update mutation that fails (rejecting the promise)
5. Monitor the isSubmitting form state, noting that it remains true
6. Note the log statements from the onFinish resolve nor reject are ever called

### Expected behavior

The promise returned by onFinish should reject and the form state should revert to isSubmitting set to false.
A simple fix would be to just remove the "return" from line 581 so that `reject` is called after `onMutationError` executes. 
(Perhaps reject should be called with the error passed in as an argument, but that may be a bigger and undesired change.)

### Screenshot

_No response_

### Desktop

_No response_

### Mobile

_No response_

### Additional Context

_No response_

### LINKED ISSUE #5702: [REFACTOR] useForm hook callbacks and mutations
We'll make some refactoring in `@refinedev/core`'s `useForm` hook to make the current codebase more readable and clean. 

While doing so there will be some changes made to the logic to make it simple and easy to follow both for the users and the maintainers.

Discussed changes is listed below but not might be more to be added;

- Resolve and reject conditions for the `onFinish` and being able to await for the mutation in pessimistic mode,
- `onMutationSuccess` and `onMutationError` will be triggered by the response of the mutation regardless of the mutation mode,
- More thorough tests both unit and e2e,
- Better integration with the devtools monitor,
- Simpler auto save logic with no separate submission block,
- Appropriate values and errors should be provided to the callbacks and promises,
- Avoiding any hanging promises (e.g. when mutation is errored out and `onMutationError` is provided).
- There might be changes required for the extended versions of the `useForm` hook in integration packages and they will also be addressed in the related PRs.

### PR DESCRIPTION
**Pull request by aliemir**
Refactored the internal logic of `useForm` to be more efficient and readable, along with few bug fixes and improvements to the `useForm` hook.

These changes are related to the issue [#5702](https://github.com/refinedev/refine/issues/5702)

- ~~Due to the confusion it causes by the name, `onFinish` is deprecated and replaced with `submit`.~~
- ~~Same also applies to `onFinishAutoSave` which is replaced with `submitAutoSave`.~~
- `onFinish` now rejects when; - `values` is not provided, - `resource` is not defined, - `id` is required but not provided.
  previously these cases were silently ignored.
- Same changes also applies to `onFinishAutoSave`.
- `onFinishAutoSave` had an issue with returning the appropriate promise after being called. This resulted in unhandled promise rejections and uncontrollable resolved promises. Now it is fixed, `onFinishAutoSave` will resolve and reject based on the response of the mutation.
- When using auto save, debounced calls will now be cancelled and the respective promises will be rejected with `"cancelled by debounce"` message. These changes might require an update to the code bases that uses `onFinishAutoSave` to handle the rejection of the promise to avoid unhandled promise rejections. 
- Combined the separated submit functions into one for sake of simplicity and consistency. (internal)
- `onFinish` rejects and resolved regardless of the `onMutationSuccess` and `onMutationError` hooks are provided or not. (Resolves [#5460](https://github.com/refinedev/refine/issues/5460))
- `meta` values were concatenated multiple times causing confusion and unexpected behavior, now it is fixed. (internal)
- Moved the `id` determination/inference logic to a separate hook. (internal)

## PR Checklist

- [x] The commit message follows our guidelines: https://refine.dev/docs/guides-concepts/contributing/#commit-convention
- [x] Related issue(s) linked
- [x] Tests for the changes have been added
- [x] Docs have been added / updated
- [x] Changesets have been added https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset

## Issues

Resolves #5702 
Fixes #5460

### CODE REVIEW

File: packages/core/src/hooks/form/index.ts [hunk 1]
```diff
@@ -1 +1,341 @@
...
+  const isCreate = action === "create";
+
+  const combinedMeta = getMeta({
+    resource,
+    meta: pickedMeta,
+  });
+
+  const idRequired = (isEdit || isClone) && Boolean(props.resource);
+  const idHandled = typeof props.id !== "undefined";
+  const idDismissed = props.queryOptions?.enabled === false;
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:112]:
```suggestion
  const isQueryDisabled = props.queryOptions?.enabled === false;
```

File: packages/core/src/hooks/form/index.ts [hunk 2]
```diff
@@ -1 +1,341 @@
...
+  const isClone = action === "clone";
+  const isCreate = action === "create";
+
+  const combinedMeta = getMeta({
+    resource,
+    meta: pickedMeta,
+  });
+
+  const idRequired = (isEdit || isClone) && Boolean(props.resource);
+  const idHandled = typeof props.id !== "undefined";
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:111]:
```suggestion
  const hasIdProp = typeof props.id !== "undefined";
```

File: packages/core/src/hooks/form/index.ts [hunk 3]
```diff
@@ -1 +1,341 @@
...
+  const idDismissed = props.queryOptions?.enabled === false;
+
+  /**
+   * When a custom resource is provided through props, `id` will not be inferred from the URL to avoid any potential faulty requests.
+   * In this case, `id` is required to be passed through props.
+   * If `id` is not handled, a warning will be thrown in development mode.
+   */
+  warnOnce(
+    idRequired && !idHandled && !idDismissed,
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:120]:
```suggestion
    idRequired && !hasIdProp && !isQueryDisabled,
```

With the suggested changes above, I think this condition becomes easier to read.

File: packages/core/src/hooks/form/index.ts [hunk 4]
```diff
@@ -1 +1,341 @@
...
+  });
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: This is a controlled unmounting effect.
+  React.useEffect(() => {
+    // After `autosaved` is set to `true`, it won't be set to `false` again.
+    // Therefore, the `invalidate` function will be called only once at the end of the hooks lifecycle.
+    return () => {
+      if (props.autoSave?.invalidateOnUnmount && autosaved) {
+        invalidate({
+          id,
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:189]: Should we check existence of `id` here?

File: packages/core/src/hooks/form/index.ts [hunk 5]
```diff
@@ -1 +1,341 @@
...
+    const onSuccessRedirect = (id?: BaseKey) => redirect(redirectAction, id);
+
+    const submissionPromise = new Promise<
+      // biome-ignore lint/suspicious/noConfusingVoidType: Void is an expected case for this promise.
+      CreateResponse<TResponse> | UpdateResponse<TResponse> | void
+    >((resolve, reject) => {
+      // Reject the mutation if the resource is not defined
+      if (!resource) return reject(missingResourceError);
+      // Reject the mutation if the `id` is not defined in edit action
+      if (isEdit && !id) return reject(missingIdError);
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:214]: I think we shouldn't reject here. AFAIk, it wasn't rejected before. And users may have a case where they update resource without an id e.g.: `api/users/settings`.

Comment (BatuhanW) [reply packages/core/src/hooks/form/index.ts:214]:
```suggestion
      if (isClone && !id) return reject(missingIdError);
```

I think we should fail if clone is missing id tho.

File: packages/core/src/hooks/form/index.ts [hunk 6]
```diff
@@ -1 +1,341 @@
...
+      if (!values) return reject(missingValuesError);
+
+      if (!isPessimistic && !isAutosave) {
+        // If the mutation mode is not pessimistic, handle the redirect immediately in an async manner
+        // `setWarnWhen` blocks the redirects until set to `false`
+        // If redirect is done before the value is properly set, it will be blocked.
+        // We're deferring the execution of the redirect to ensure that the value is set properly.
+        deferExecution(() => onSuccessRedirect());
+        // Resolve the promise immediately
+        resolve();
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:225]: We can resolve with mutation variables for optimistic update.

File: packages/core/src/hooks/form/index.ts (hunk 3 above)
Comment (BatuhanW) [reply packages/core/src/hooks/form/index.ts:120]: I think we can also use `id` variable here, coming from `useId` hook.

File: packages/core/src/hooks/form/index.ts [hunk 7]
```diff
@@ -1 +1,341 @@
...
+
+  const {
+    resource,
+    action: routeAction,
+    identifier,
+  } = useResource(props.resource);
+
+  const [autosaved, setAutosaved] = React.useState(false);
+
+  const [id, setId] = useId({ id: props.id, resource: props.resource });
```
Comment (BatuhanW) [inline packages/core/src/hooks/form/index.ts:90]:
```suggestion
  const [id, setId] = useId({ id: props.id, resource: props.resource });
```

Do we handle the case, `resource` string is not the name of the string, but it's `identifier` of the resource?

**Review by omeraplak — APPROVED** [review]

**Pull request merged**